packages/data-provider/src/utils/submittedDataUtils.ts

```diff
@@ -16,15 +17,74 @@ import {
 import { BaseDependencies } from '../config/config.js';
 import {
 	DataRecordReference,
-	MERGE_REFERENCE_TYPE,
 	type GroupedDataSubmission,
+	MERGE_REFERENCE_TYPE,
+	type MutableDataDiff,
 	type SubmittedDataResponse,
 } from './types.js';
 
 const utils = (dependencies: BaseDependencies) => {
 	const LOG_MODULE = 'SUBMITTED_DATA_UTILS';
 	const { logger } = dependencies;
 	return {
+		/**
+		 * Compares two `DataRecord` objects and returns the differences between them.
+		 * @param oldRecord The original `DataRecord` object to compare.
+		 * @param newRecord The new `DataRecord` object to compare against the original.
+		 * @returns An object of type `DataDiff` containing the differences between `oldRecord` and `newRecord`.
+		 * 	The differing values are recorded with the `old` object containing the values
+		 * 	from `oldRecord` and the `new` object containing the corresponding values from `newRecord`.
+		 */
+		computeDataDiff: (oldRecord: DataRecord | null, newRecord: DataRecord | null): DataDiff => {
+			const diff: MutableDataDiff = { old: {}, new: {} };
+
+			if (!oldRecord && !newRecord) {
+				// Both records are null, no differences to return
+				return diff;
+			}
+
+			if (!oldRecord) {
+				// oldRecord is null, all keys in newRecord are new
+				for (const key in newRecord) {
+					if (Object.prototype.hasOwnProperty.call(newRecord, key)) {
+						diff.new[key] = newRecord[key];
+					}
+				}
```

Comment:
I think I know why you included this if check, but for the record and to make sure this is needed...

Why are we checking that the record has the key, when you just got the key from the record? Is there some subset of properties that are included in the object iterator but are not included in `hasOwnProperty` for the record?

Reply:
This is a check to ensure that the key being iterated over in the `for...in` loop actually belongs to the newRecord object itself and is not an inherited property from its prototype like `toString` or others